zipper/protocols/irondb/irondb_group.go

```diff
@@ -0,0 +1,708 @@
+	return nil, nil, types.ErrNotSupportedByBackend
+}
+
+func (c *IronDBGroup) List(ctx context.Context) (*protov3.ListMetricsResponse, *types.Stats, merry.Error) {
+	return nil, nil, types.ErrNotImplementedYet
+}
+
+func (c *IronDBGroup) Stats(ctx context.Context) (*protov3.MetricDetailsResponse, *types.Stats, merry.Error) {
+	return nil, nil, types.ErrNotSupportedByBackend
+}
+
+func (c *IronDBGroup) doTagQuery(ctx context.Context, isTagName bool, query string, limit int64) ([]string, merry.Error) {
+	logger := c.logger
+	params := make(map[string][]string)
+	var result []string
+	var target string
+	var tag_category string
+
+	// decoding query
+	queryDecoded, _ := url.QueryUnescape(query)
+	querySplit := strings.Split(queryDecoded, "&")
+	for _, qvRaw := range querySplit {
+		idx := strings.Index(qvRaw, "=")
+		//no parameters passed
+		if idx < 1 {
+			continue
+		}
+		k := qvRaw[:idx]
+		v := qvRaw[idx+1:]
+		if v2, ok := params[qvRaw[:idx]]; !ok {
+			params[k] = []string{v}
+		} else {
+			v2 = append(v2, v)
+			params[k] = v2
+		}
+	}
+	logger.Debug("doTagQuery",
+		zap.Any("query", queryDecoded),
+		zap.Bool("isTagName", isTagName),
+		zap.Any("params", params),
+		zap.Int64("limit", limit),
+	)
+
+	// default target - all Graphite metrics
+	target = `__name=~.*`
+	// but use joined expr value if present
+	if len(params["expr"]) > 0 {
+		target = strings.Join(params["expr"], ",")
+	}
+	if isTagName {
+		logger = logger.With(zap.String("type", "tagName"))
+	} else {
+		logger = logger.With(zap.String("type", "tagValues"))
+		// get tag category (if present and we're looking for values instead categories)
+		if tag, ok := params["tag"]; ok {
+			tag_category = tag[0]
+		} else {
+			return []string{}, types.ErrNoTagSpecified
+		}
+	}
+
+	logger.Debug("sending GraphiteFindTags request to irondb",
+		zap.Int64("accountID", c.accountID),
+		zap.String("prefix", c.graphite_prefix),
+		zap.String("target", target),
+	)
+	tag_result, err := c.client.GraphiteFindTags(c.accountID, c.graphite_prefix, target, nil)
+	if err != nil {
+		return []string{}, merry.New("request returned an error").WithValue("error", err)
+	}
+	logger.Debug("got GraphiteFindTags result from irondb",
+		zap.String("target", target),
+		zap.Any("tag_result", tag_result),
+	)
+	for _, metric := range tag_result {
+		// if metric name contain tags
+		if strings.Contains(metric.Name, ";") {
+			namex := strings.Split(metric.Name, ";")
+			for i := 1; i < len(namex); i++ {
+				tag := strings.SplitN(namex[i], "=", 2)
+				// tag[0] is tag category and tag[1] is tag value
+				if isTagName {
+					// if filtering by tag prefix
+					if v, ok := params["tagPrefix"]; ok {
+						// and prefix match
+						if strings.HasPrefix(tag[0], v[0]) {
+							// append tag name to result
+							result = append(result, tag[0])
+						}
+					} else {
+						// if not filtering by tag prefix - append all tags
+						result = append(result, tag[0])
+					}
+				} else {
```

Comment:
This else seems to be almost identical to the if itself. I think it's a good candidate to unification, e.x.
```
tagIdx := 1
if isTagName {
  tagIdx = 1
}
<...>
```

And instead of `if tag[0] == tag_category` do something like
```
if tagIdx == 0 || (tag[0] == tagCategory && tagIdx == 1) {
  result = append(result, tag[tagIdx])
}
```

Reply:
Yes, it looks optimizable, but when I tried I quickly getting completely non-readable or not properly working code... will try to optimize it tomorrow

Reply:
Completely reworked in https://github.com/go-graphite/carbonapi/pull/626/commits/fb45ebbc840a6100f0df513562d91ed5373643ce